### DISCUSSION TITLE
Export the FEM matrix

### DISCUSSION THREAD

**Discussion by TorgeW** [simnibs/simnibs#213] [9. Results and analysis] [answered]
Hello, I am using SimNIBS for TMS simulations. For a given geometry and stimulation I would like to export the FEM matrix (i.e. system matrix) and the right-hand-side (i.e. load vector), is this possible? thank you very much.

@lucadirienzo (Continue your question and the discussion here please)

**Comment by jduemose** [comment, marked as answer]
hi there

If you look in `simnibs.simulation.fem` there is a class called `TMSFEM` which we use for the TMS simulations. Below is an example (**completely untested**) of how one could obtain the system matrix and RHS

        import numpy as np

        import simnibs.mesh_tools.mesh_io
        from simnibs.simulation.fem import _get_da_dt_from_coil, TMSFEM
        from simnibs.simulation.sim_struct import POSITION
        from simnibs.utils.cond_utils import cond2elmdata, standard_cond

        #  ========= INPUTS =========
        filename_mesh = "/path/to/subid.msh"
        filename_coil = "/path/to/coilfile"

        # Coil position info (here example) see help(simnibs.simulation.sim_struct.POSITION)
        tmspos = POSITION()
        tmspos.name = "name of simulation"
        tmspos.didt = 1.0 
        tmspos.centre = [0,1,2] 
        tmspos.pos_ydir = [0,1,2] 
        # tmspos.distance = # distance to head
        
        # or
        # tmspos.matsimnibs = ... # if you have that already

        # ========= CALCULATE =========

        mesh = simnibs.mesh_tools.mesh_io.read_msh(filename_mesh)

        if tmspos.matsimnibs is None:
            _ = tmspos.calc_matsimnibs(mesh)

        # get conductivities per element
        std_cond = standard_cond()
        cond_list = [c.value for c in std_cond]
        cond = cond2elmdata(mesh, cond_list)

        # setup da/dt
        dadt = _get_da_dt_from_coil(
            filename_coil, mesh, np.atleast_1d(tmspos.didt), tmspos.matsimnibs
        )

        # Now assemble system matrix and compute RHS
        system = TMSFEM(mesh, cond)
        rhs = system.assemble_rhs(dadt)
        stiffness_matrix = system.A
        # v = system.solve(b) # solve

**Reply by TorgeW** [reply]
@lucadirienzo

**Reply by lucadirienzo** [reply]
Thank you very much for your help. Apparently there is no function _get_da_dt_from_coil in fem.py.
However, in fem.py they refer to dadt: simnibs.msh.mesh_io.NodeData (optional)
dA/dt at the nodes for TMS simulations
Do you know how to extract dadt to use it in system.assemble_rhs(dadt)?

Further there seems to be no module named 'simnibs.utils.cond_utils'

Do you know something more about it?

Thanks a lot!

**Reply by jduemose** [reply]
My mistake I was looking at a development branch which is why the code was slightly different. I apologize. However, if you go to the release page there should be a pre-release of version 4.1 which should match the above description. Try to download that and see if it works.

The data of a `NodeData` class is stored in the `.value` attribute.

Best,
Jesper

**Comment by lucadirienzo** [comment]
Thanks again for the helpful information. Finally I got it to run and exportet the stiffness_matrix and the rhs, 
bye,
Luca
